### ISSUE TITLE
Store.findRecords failing on flutter with dart 2

### ISSUE THREAD

**Issue by andreiboyanov** [tekartik/sembast.dart#5]
Thanks for the great package !

My problem: the findRecords method is failing on Flutter with Dart 2.

How to reproduce
----------------------
1. Create a new Flutter project
2. Use the following pubspec.yaml:

```
name: sembast_flutter_tests
description: Test of the sembast package under Flutter

dependencies:
  flutter:
    sdk: flutter
  sembast: ^1.7.0
  path_provider: any
  cupertino_icons: ^0.1.2

dev_dependencies:
  flutter_test:
    sdk: flutter

flutter:
  uses-material-design: true
```
3. Put the following code in lib/main.dart:

```
import "package:sembast/sembast_io.dart";
import "package:sembast/sembast.dart";
import 'package:path/path.dart';
import 'package:path_provider/path_provider.dart';

void main() {
  getApplicationDocumentsDirectory().then((dataDir) {
    final dbPath = join(dataDir.path, "test.db");
    ioDatabaseFactory.openDatabase(dbPath).then((db) {
      final store = db.getStore("test store");
      final record = new Record(
        store,
        {"field1": "value for field 1"},
        1
      );
      db.putRecord(record).then((record) {
        Store store = db.getStore("test store");
        final finder = new Finder(limit: 1);
        store.findRecords(finder).then((records) {
          print(records);
        });
      });
    });
  });
}
```
4. Execute `flutter run`. I have the following output:
```
Using hardware rendering with device Android SDK built for x86. If you get graphics artifacts, consider enabling software rendering with "--enable-software-rendering".
Launching lib/main.dart on Android SDK built for x86 in debug mode...
Initializing gradle...                                       0.9s
Resolving dependencies...                                    0.8s
Running 'gradlew assembleDebug'...                           9.7s
Built build/app/outputs/apk/debug/app-debug.apk.
Installing build/app/outputs/apk/app.apk...                  1.4s
I/FlutterActivityDelegate(13860): onResume setting current activity to this
I/flutter (13860): test store
E/flutter (13860): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
E/flutter (13860): type 'Future<dynamic>' is not a subtype of type 'Future<List<Record>>' in type cast
E/flutter (13860): #0      Object._as (dart:core/runtime/libobject_patch.dart:67:25)
E/flutter (13860): #1      SembastStore.findRecords (package:sembast/src/store_impl.dart:115:8)
E/flutter (13860): #2      main.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///home/andrei/src/sembast_flutter_tests/lib/main.dart:20:15)
E/flutter (13860): #3      _RootZone.runUnary (dart:async/zone.dart:1381:54)
E/flutter (13860): #4      _FutureListener.handleValue (dart:async/future_impl.dart:129:18)
E/flutter (13860): #5      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:633:45)
E/flutter (13860): #6      Future._propagateToListeners (dart:async/future_impl.dart:662:32)
E/flutter (13860): #7      Future._complete (dart:async/future_impl.dart:467:7)
E/flutter (13860): #8      _SyncCompleter.complete (dart:async/future_impl.dart:51:12)
E/flutter (13860): #9      SembastDatabase.inTransaction.<anonymous closure>.<anonymous closure> (package:sembast/src/database_impl.dart)
E/flutter (13860): <asynchronous suspension>
E/flutter (13860): #10     SynchronizedLock._runTask.<anonymous closure>.<anonymous closure> (package:synchronized/src/synchronized_impl.dart:189:29)
E/flutter (13860): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter (13860): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter (13860): #13     runZoned (dart:async/zone.dart:1501:17)
E/flutter (13860): #14     SynchronizedLock._runTask.<anonymous closure> (package:synchronized/src/synchronized_impl.dart:187:14)
E/flutter (13860): #15     new Future.sync (dart:async/future.dart:222:31)
E/flutter (13860): #16     SynchronizedLock._runTask (package:synchronized/src/synchronized_impl.dart:186:16)
E/flutter (13860): #17     LockBase._createAndRunTask.run (package:synchronized/src/synchronized_impl.dart:61:14)
E/flutter (13860): #18     LockBase._createAndRunTask (package:synchronized/src/synchronized_impl.dart:71:17)
E/flutter (13860): #19     SynchronizedLock.synchronized (package:synchronized/src/synchronized_impl.dart:222:12)
E/flutter (13860): #20     SembastDatabase.inTransaction.<anonymous closure> (package:sembast/src/database_impl.dart:89:19)
E/flutter (13860): #21     new Future.<anonymous closure> (dart:async/future.dart:174:37)
E/flutter (13860): #22     Timer._createTimer.<anonymous closure> (dart:async/runtime/libtimer_patch.dart:21:15)
E/flutter (13860): #23     _Timer._runTimers (dart:isolate/runtime/libtimer_impl.dart:382:19)
E/flutter (13860): #24     _Timer._handleMessage (dart:isolate/runtime/libtimer_impl.dart:416:5)
E/flutter (13860): #25     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:165:12)
Syncing files to device Android SDK built for x86...         0.9s

🔥  To hot reload your app on the fly, press "r". To restart the app entirely, press "R".
An Observatory debugger and profiler on Android SDK built for x86 is available at: http://127.0.0.1:8111/
For a more detailed help message, press "h". To quit, press "q".
```

`flutter doctor -v` shows the following:
```
[✓] Flutter (Channel master, v0.4.5-pre.52, on Linux, locale en_US.UTF-8)
    • Flutter version 0.4.5-pre.52 at /home/andrei/opt/flutter
    • Framework revision 8717a5e6f7 (2 days ago), 2018-05-19 03:19:35 +0300
    • Engine revision 1179c38a42
    • Dart version 2.0.0-dev.55.0.flutter-43635d3372

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /home/andrei/Android/Sdk
    • Android NDK at /home/andrei/Android/Sdk/ndk-bundle
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = /home/andrei/Android/Sdk
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] Android Studio (version 3.1)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 24.2.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] Connected devices (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.0.0 (API 26) (emulator)

• No issues found!
```

What I really don't understand is why the same code is running without problem under Dart VM with the same dart version as above...

Thanks for the help.

**Comment by alextekartik** [comment]
Thanks really much for your complete detailed report and indeed the only way to reproduce it was in flutter. There were some casting issue and some timing issue on flutter that are new in dev.55. I think your issue should be fixed in the version 1.8.0. However lot me know how it goes. I'm doing more flutter testing on my side too

**Comment by andreiboyanov** [comment]
That' s great, thanks ! 1.8.0 fixes my issue indeed. I'll let you know if I find it anywhere else.
